### PR TITLE
Allow more methods to take interned arguments

### PR DESCRIPTION
**Pull request by mejrs**
- [x] I have yet to add some docs to these methods
- [x] Changelog

### CODE REVIEW

File: src/types/string.rs
Comment (mejrs) [inline src/types/string.rs:307]: This is a (mild) breaking change because of inference errors, but not too bad.

Comment (adamreichold) [reply src/types/string.rs:307]: Just to make sure I understand things correctly: The comment is informational as the other changes are breaking anyway and this is therefore targeting 0.17.0?

Comment (birkenfeld) [reply src/types/string.rs:307]: Is the previous instance into PyObject even useful anymore?

**Comment by birkenfeld** [comment]
Very nice!

File: src/types/string.rs
Comment (adamreichold) [reply src/types/string.rs:307]: > Is the previous instance into PyObject even useful anymore?

I think so as `Py<PyString>` does not automatically "decay" into `Py<PyAny>` and the existing impl might be used in generic context to produce `PyObject` for stashing away.

Comment (birkenfeld) [reply src/types/string.rs:307]: Easy to get PyObject from a Py though.

Comment (adamreichold) [reply src/types/string.rs:307]: > Easy to get PyObject from a Py though.

I think that in generic context, one needs to be able to express this in a single trait bound, e.g. calling `into_py` twice or `into_py` followed by `into` to go from `T` to `Py<S>` to `PyObject` is not necessarily the same as going from `T` to `PyObject`.

Comment (mejrs) [reply src/types/string.rs:307]: > Just to make sure I understand things correctly: The comment is informational as the other changes are breaking anyway and this is therefore targeting 0.17.0?

Mainly it means that things like `let thing = "thing".into_py(py);` stop compiling

Comment (davidhewitt) [reply src/types/string.rs:307]: To avoid making lots of annoying inference errors we could potentially combine this PR with #2316, which would enable us to select on types which have `IntoPyObject<Target = PyString>` instead of `IntoPy<Py<PyString>>`.

File: src/types/any.rs
Comment (adamreichold) [inline src/types/any.rs:107]: As things are, this would leave us with two different kinds of bounds for names, either `N: ToPyObject` or `N: IntoPy<Py<PyString>>`. Would it make sense to unify them?

File: src/types/any.rs
Comment (adamreichold) [inline src/types/any.rs:551]: Maybe this would be a good place to showcase using the `intern!` macro to make it easier for the reader of the documentation to understand how it is supposed to be used?

**Review by adamreichold — APPROVED** [review]
Looks good to me.

One remaining concern that I see is that we have made several methods "trivially generic" which were not generic before. Should we aim to reduce code bloat and factor out the non-generic parts into inner help functions? On the other hand, the functions in question do not appear to be huge, so I am unsure whether this would be wort it.

File: src/types/any.rs
Comment (mejrs) [reply src/types/any.rs:107]: I looked into it and I think attribute names don't need to be strings - anything will do as long as it inherits from `str`.

I would be happy to limit it to `IntoPy<Py<PyString>>`, but - are people relying on this behaviour?

For example:
```rust
use pyo3::prelude::*;

#[pyfunction]
fn getattr(item: &PyAny, arg: &PyAny) -> PyResult<()> {
    item.getattr(arg)?;
    Ok(())
}

/// This module is implemented in Rust.
#[pymodule]
fn my_module(py: Python<'_>, m: &PyModule) -> PyResult<()> {
    m.add_function(wrap_pyfunction!(getattr, m)?)?;
    Ok(())
}
```

```python
import my_module

class A:
    def __getattr__(self, attr):
        print(attr)
        print(attr.__class__)
        return "foo"

a = A()

class C(str):
    pass

my_module.getattr(a, "field")
my_module.getattr(a, C())
```

Comment (davidhewitt) [reply src/types/any.rs:107]: I think `IntoPy<Py<PyString>>` makes sense for attribute names. Edge cases like this don't seem likely to me, and I think with a bit of type wrangling (e.g. change `arg: &PyAny` to `arg: &PyString`) things should still work.

**Comment by mejrs** [comment]
> Should we aim to reduce code bloat and factor out the non-generic parts into inner help functions?

They are pretty small, I think it's fine.

> Would it make sense to unify them?

Done 👍

**Comment by davidhewitt** [comment]
Looks good to me too. The only question I have outstanding is whether we want to merge this now or wait until we can also merge with #2316 so that we don't cause users to go through extra churn of type inference failures in the interim.

(FWIW I have a bad feeling that #2316 gets a little upset with some of the more gnarly `#[pyproto]` code, so that might be blocked until 0.18 when we can kill the feature off entirely.)

**Comment by mejrs** [comment]
I'm feeling that what we're doing with traits is a bit ad-hoc.

Maybe we need to sit down and think about what we don't like, what we would like, and what a good path to getting there is.

File: src/types/any.rs
Comment (mejrs) [reply src/types/any.rs]: Done

**Comment by davidhewitt** [comment]
> I'm feeling that what we're doing with traits is a bit ad-hoc.
> 
> Maybe we need to sit down and think about what we don't like, what we would like, and what a good path to getting there is.

I agree very much with this sentiment. I'll write my current backlog of ideas up in a separate issue as a starting point.

As for this PR... looks good to me! The concern I have with merging before we make a decision on #2316 is that this may cause users to have to add `: PyObject` annotations to keep existing `.into_py()` uses compiling. If we then merge #2316 these annotations may then become type errors. (I guess #2316 will probably break a bit of existing code too, so maybe that's not actually a reason to hold back this PR?) What's your opinion on this?

File: src/instance.rs
Comment (mejrs) [inline src/instance.rs:670]: Sorry this is a bit messy since resolving the merge conflict.

How do you feel about:

```rust
pyo3::ffi::Something(name.to_object(py).as_ptr());
```

versus 

```rust
let ptr =  name.to_object(py).into_ptr();
pyo3:;ffi::Something(ptr);
ffi::Py_DECREF(ptr);
```
(or something else?)

Personally I prefer the latter. IMO the former is error prone and gives a bad example.

**Comment by mejrs** [comment]
>  What's your opinion on this?

I think that's fine, it's breaking either way.

File: src/instance.rs
Comment (davidhewitt) [inline src/instance.rs:719]: Exactly, like this is how I think we should be making use of `Py` to avoid leaking memory 👍

File: src/types/any.rs
Comment (davidhewitt) [inline src/types/any.rs:188]: e.g. here there's no need for `XDECREF` instead of `DECREF` because it's got to be a non-null pointer (it came via `to_object` call).

**Review by davidhewitt — COMMENTED** [review]
Ok, let's head towards merging this in 0.17 and maybe thinking about `IntoPyObject` for 0.18. I'll try to write up my thoughts on traits tomorrow.

Got to confess I'm really not a fan of calling `Py_DECREF` or `Py_XDECREF` whenever we can just let `Py` do it in `Drop` implementation.

File: src/instance.rs
Comment (mejrs) [reply src/instance.rs:670]: > Why do you feel the former is error prone? I think the second is more error prone because missing the `Py_DECREF` call is leaking memory just like missing `free()` in C.

I'm mostly worried about someone reading it and and writing it slightly differently:
```rust
let ptr = name.to_object(py).as_ptr();
pyo3::ffi::Something(ptr);
```
which produces a dangling pointer, as the python object is dropped at the end of the statement (I should add documentation about this).

> Perhaps a middle ground is to make it obvious we're using RAII through `Py`?
> 
> ```rust
> let name: PyObject = name.to_object(py);
> pyo3::ffi::Something(name.as_ptr());
> ```

Sure 👍

Comment (davidhewitt) [reply src/instance.rs:670]: > 'm mostly worried about someone reading it and and writing it slightly differently:

This is an excellent point which I'd totally missed. IMO the middle ground above is probably the only sensible way to arrange this then. If you like, I'll submit a PR to fix the whole codebase to use this pattern once this PR merged? That way you can just worry about the modified functions here without another merge conflict.

**Review by davidhewitt — COMMENTED** [review]
👍 Looks great, thanks very much for all the tidy ups!

A final thought: as we're committing to merging this and producing extra inference errors in PyO3 0.17, can we please add a section to `migration.md` to explain this to users?

**Comment by mejrs** [comment]
> A final thought: as we're committing to merging this and producing extra inference errors in PyO3 0.17, can we please add a section to `migration.md` to explain this to users?

Sure.

I'm also concerned about some missing codecov here, so I want to add a couple of tests as well.

**Comment by mejrs** [comment]
Should be done now :)

**Comment by davidhewitt** [comment]
Thanks for merging, looks great, sorry I had a busy last few days and didn't get around to looking at this sooner. 👍

**Pull request merged**
